lib/select.ts

```diff
@@ -0,0 +1,201 @@
+    })
+  );
+  return container;
+}
+
+/**
+ * Creates an instance of Menu intended to mimic the behavior of the select HTML element.
+ *
+ * Should always be created using select(), which adds the select button and associated logic.
+ */
+class Select<T> extends BaseMenu {
+  private _selectRows: HTMLElement[] = Array.from(this.content.children) as HTMLElement[];
+  private _keyState: SelectKeyState<string> = new SelectKeyState(() =>
+    this._selectRows.map(_elem => ({
+      label: _elem.textContent || "",
+      value: _elem.textContent || "",
+      disabled: _elem.classList.contains('disabled')
+    }))
+  );
+
+  constructor(ctl: IOpenController, items: DomElementArg[], options: ISelectOptions = {}) {
+    super(ctl, items, options);
+
+    // On keydown, search for the first element with a matching label.
+    onElem(this.content, 'keydown', (ev) => {
+      const sel = this._keyState.add(ev.key);
+      if (sel) { this._selectRow(sel.label); }
+    });
+
+    // If an initial value is given, use it, otherwise select the first element.
+    const init = options.selectLabelOnOpen ? options.selectLabelOnOpen() : null;
+    setTimeout(() => (init ? this._selectRow(init) : this.nextIndex()), 0);
+  }
+
+  private _selectRow(label: string): void {
+    const elem = this._selectRows.find(_elem => _elem.textContent === label);
+    return this.setSelected(elem || null);
+  }
+}
+
+
+/**
+ * Maintains the state of the user's keyed-in value when searching in the select element.
+ */
+class SelectKeyState<T> {
+  private _term: string = "";
+  private _cycleIndex: number = 0;
+  private _timeoutId: NodeJS.Timeout|null = null;
+
+  constructor(private _itemCallback: () => IOption<T>[]) {}
+
+  // Adds a character to the search term. Returns the latest first match of the items, or null
+  // if no items match.
+  public add(char: string): IOptionFull<T>|null {
+    // Filter out any keys that are not a single character. Make all searching case-insensitive.
+    if (char.length > 1) { return null; }
+    char = char.toLowerCase();
+
+    // Clear the search term after a timeout. Any new entry should reset the timeout.
+    if (this._timeoutId) { clearTimeout(this._timeoutId); }
+    this._timeoutId = setTimeout(() => { this._term = ""; }, 1000);
+
+    // Add character to the search term and look for a match. If a match is found, update the
+    // observable value.
+    if (this._term.length === 1 && char === this._term[0]) {
+      // If the same character is pressed repeatedly, cycle through all options starting with
+      // that character.
+      this._cycleIndex += 1;
+    } else {
+      // Add character to the search term and reset the cycle search index.
+      this._term += char;
+      this._cycleIndex = 0;
+    }
+    const matches = this._itemCallback().filter(_item => {
+      _item = getOptionFull(_item);
+      return !_item.disabled && _item.label.toLowerCase().slice(0, this._term.length) === this._term;
```

Comment:
Perhaps `_item.label.toLowerCase().startsWith(this._term)`? (`slice` is more expensive -- it creates a new string, and does that for each option).